### DISCUSSION TITLE
asdf install ruby latest installing outdated openssl version, failing

### DISCUSSION THREAD

**Discussion by mgibson323** [rbenv/ruby-build#2237] [Build failures]
### Steps to reproduce the behavior

On MacOS 13.4.1 (ARM64) with `asdf` installed via `brew`, and openssl version `3.1.2`, attempt to install `ruby latest`

```
$ asdf version
v0.12.0
```

```
$ openssl version
OpenSSL 3.1.2 1 Aug 2023 (Library: OpenSSL 3.1.2 1 Aug 2023)
```

```
$ asdf install ruby latest
```

The installation fails with other ruby versions besides `latest`, as well. 

I suspect the openssl version is causing my installation to fail. 

Note: I am not very familiar with Ruby or using these tools, but I want to install it to use it for some other tools I need for work. I was directed to this repo by [this comment](https://github.com/asdf-vm/asdf/issues/934#issuecomment-845411732).

### Expected vs. actual behavior

Expected: ISBAT install Ruby using `asdf` so I can use a different version aside from the MacOS default. 
Actual: Installation via `asdf` fails

### Logs

```
$ asdf install ruby latest
To follow progress, use 'tail -f /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230804125921.3441.log' or pass --verbose
Downloading openssl-3.1.1.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/b3aa61334233b852b63ddb048df181177c2c659eb9d4376008118f9c08d07674
Installing openssl-3.1.1...

BUILD FAILED (macOS 13.4.1 using ruby-build 20230717)

Inspect or clean up the working tree at /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230804125921.3441.YNnJS2
Results logged to /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230804125921.3441.log

Last 10 log lines:
  "_main", referenced from:
     implicit entry/start for main executable
  "_test_info", referenced from:
      _test_cha_cha_internal in chacha_internal_test-bin-chacha_internal_test.o
  "_test_uchar_eq", referenced from:
      _test_cha_cha_internal in chacha_internal_test-bin-chacha_internal_test.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [test/chacha_internal_test] Error 1
make: *** [build_sw] Error 2
```

**Comment by mgibson323** [comment]
additional info: 

```
$ asdf info
OS:
Darwin gitlab.local 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000 arm64

SHELL:
zsh 5.9 (x86_64-apple-darwin22.0)

BASH VERSION:
5.2.15(1)-release

ASDF VERSION:
v0.12.0

ASDF INTERNAL VARIABLES:
ASDF_DEFAULT_TOOL_VERSIONS_FILENAME=.tool-versions
ASDF_DATA_DIR=/Users/<user>/.asdf
ASDF_DIR=/opt/homebrew/Cellar/asdf/0.12.0/libexec
ASDF_CONFIG_FILE=/Users/<user>/.asdfrc

ASDF INSTALLED PLUGINS:
ruby                         https://github.com/asdf-vm/asdf-ruby.git master 8cb9721
```

**Comment by mislav** [comment]
I have no idea what is causing this failure, but you could avoid having to compile OpenSSL by linking to an existing one.

You could install an OpenSSL version with Homebrew, which is generally faster due to Homebrew's precompiled bottles: `brew install openssl@1.1`

Then you can instruct ruby-build to link to that:
```sh
RUBY_CONFIGURE_OPTS="--disable-install-doc --with-openssl-dir=$(brew --prefix openssl@1.1)"
```

**Reply by mgibson323** [reply]
I do have a Brew version of OpenSSL installed. 

```
$ brew ls --versions openssl
openssl@3 3.1.2
```

Thanks, I'll try that.

**Reply by mgibson323** [reply]
Update: 

```
$ RUBY_CONFIGURE_OPTS="--disable-install-doc --with-openssl-dir=$(brew --prefix openssl@3)"
```

```
$ asdf install ruby latest
To follow progress, use 'tail -f /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230808094930.62467.log' or pass --verbose
Downloading openssl-3.1.1.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/b3aa61334233b852b63ddb048df181177c2c659eb9d4376008118f9c08d07674
Installing openssl-3.1.1...

BUILD FAILED (macOS 13.4.1 using ruby-build 20230717)

Inspect or clean up the working tree at /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230808094930.62467.nDuHlx
Results logged to /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230808094930.62467.log

Last 10 log lines:
  "_main", referenced from:
     implicit entry/start for main executable
  "_test_info", referenced from:
      _test_cha_cha_internal in chacha_internal_test-bin-chacha_internal_test.o
  "_test_uchar_eq", referenced from:
      _test_cha_cha_internal in chacha_internal_test-bin-chacha_internal_test.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [test/chacha_internal_test] Error 1
make: *** [build_sw] Error 2
```
I'm not sure why it's still trying to install `openssl-3.1.1`. WDYT? Thanks again.

**Reply by mislav** [reply]
Hi, if you were going to set an environment variable, either set it per-invocation:
```sh
RUBY_CONFIGURE_OPTS="..." asdf install ...
```
or export it to all subsequent child processes of the shell:
```sh
export RUBY_CONFIGURE_OPTS="..."
asdf install ...
```
but this won't work:
```sh
# wrong:
RUBY_CONFIGURE_OPTS="..."
asdf install ...
```
because the variable that you've set will only take effect in the current shell but won't be exported to any of that shell's sub-processes (meaning `asdf` will never receive it).

**Reply by mgibson323** [reply]
Thanks! That makes sense, my apologies for noob-ness. 

The installation is now failing differently, so I am not sure if it makes sense to keep this open. 

```
$ RUBY_CONFIGURE_OPTS="--disable-install-doc --with-openssl-dir=$(brew --prefix openssl@3)" asdf install ruby latest
To follow progress, use 'tail -f /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230809124226.49767.log' or pass --verbose
Downloading ruby-3.2.2.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.2.tar.gz
Installing ruby-3.2.2...
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.5 using ruby-build 20230717)

Inspect or clean up the working tree at /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230809124226.49767.XbjUD4
Results logged to /var/folders/89/vrmgx5c91fl4dbdf_ydq2y4w0000gn/T/ruby-build.20230809124226.49767.log

Last 10 log lines:
compiling digest.c
compiling date_core.c
linking shared-object coverage.bundle
compiling date_parse.c
compiling date_strftime.c
compiling date_strptime.c
linking shared-object digest.bundle
linking shared-object -test-/string.bundle
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```
I can open a different issue or I can share more from the log here, please let me know which is appropriate. 

Thank you!

**Reply by mislav** [reply]
You can share more of the log file here; thank you. You can also paste the whole thing as a gist link, if you cannot isolate the error message yourself.

I wonder if the Ruby compilation problem is somehow related to openssl compilation problem. As a precaution, make sure that your Homebrew installation wasn't set up in `amd64` compatibility mode (some people did this with their M1 machines because either Homebrew or various software wasn't compatible with `arm64`) and that you since forgot about it. I don't know how exactly to check for this, however. You can start with `type brew` to rule out the possibility that `brew` is a shell alias for something else.

**Reply by mgibson323** [reply]
Thanks!

I did initially have Brew set up for `amd64` compatibility, but I've since used [this method](https://earthly.dev/blog/homebrew-on-m1/) to switch it all to `arm64`. 

Here is the gist containing the full log: https://gist.github.com/mgibson323/6b295a5f346f49d8f3c696edad755fa6

**Reply by jwalinrami** [reply]
@mislav , I am also trying to install ruby on exact same machine, with same asdf and openssl version already installed along with latest brew as @mgibson323 and having BUILD FAILURE. I had also re-installed Xcode 14.3.1 and Xcode command tools (14.3.1)  prior to below attempt. Also - not sure if that was required but I have added my apple developer account in the xcode 'settings' as well if that were to help.

**************
asdf install ruby latest
To follow progress, use 'tail -f /var/folders/6n/h3k8tctx53zblr2d5s9yw1vctp0f93/T/ruby-build.20230811112630.33605.log' or pass --verbose
Downloading openssl-3.1.1.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/b3aa61334233b852b63ddb048df181177c2c659eb9d4376008118f9c08d07674
Installing openssl-3.1.1...

BUILD FAILED (macOS 13.4.1 using ruby-build 20230717)

Inspect or clean up the working tree at /var/folders/6n/h3k8tctx53zblr2d5s9yw1vctp0f93/T/ruby-build.20230811112630.33605.fpJCJh
Results logged to /var/folders/6n/h3k8tctx53zblr2d5s9yw1vctp0f93/T/ruby-build.20230811112630.33605.log

Last 10 log lines:
make[1]: *** [crypto/ffc/libcrypto-lib-ffc_backend.o] Killed: 9
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [crypto/evp/libcrypto-lib-p_verify.o] Killed: 9
make[1]: *** Deleting file `crypto/evp/libcrypto-lib-p_verify.o'
make[1]: *** [crypto/evp/libcrypto-lib-p_sign.o] Killed: 9
make[1]: *** Deleting file `crypto/evp/libcrypto-lib-p_sign.o'
/bin/sh: line 1: 38315 Killed: 9               mv crypto/evp/libcrypto-lib-p_lib.d.tmp crypto/evp/libcrypto-lib-p_lib.d
make[1]: *** [crypto/evp/libcrypto-lib-p_lib.o] Error 137
make[1]: *** [crypto/ffc/libcrypto-lib-ffc_dh.o] Killed: 9
make: *** [build_sw] Error 2
**************

And here is gist of my logs for reference, 
https://gist.github.com/jwalinrami/fd3c28d43b50b61150abafc7e2c6a2f4

**Reply by mislav** [reply]
@mgibson323 This error coming from `bigdecimal` sounds familiar:
```
./missing.h:139:1: error: static declaration of 'rb_rational_den' follows non-static declaration
```

Here is the [Ruby bug report](https://bugs.ruby-lang.org/issues/17486). I don't know why you now experience this with a much newer Ruby version, nor what causes the original issue. One person reports that linking Homebrew's `coreutils` formula into your PATH may be the culprit. For more information, you can inspect `ext/bigdecimal/mkmf.log` in your build directory.

**Comment by jwalinrami** [comment]
@mgibson323  @mislav Please ignore my report above as I was able to resolve the issue with help from our mgr who knows Ruby well and here is what we did, 

1. Uninstall and Reinstall Xcode and Xcode command line tools (14.3.1)

2. Set .zprofile file as below
****************************
eval "$(/opt/homebrew/bin/brew shellenv)"

export RUBY_CONFIGURE_OPTS="--with-libyaml-dir=$(brew --prefix libyaml) --with-openssl-dir=$(brew --prefix openssl@3) --disable-install-doc"

export PATH="/opt/homebrew/opt/openssl@3/bin:$PATH"
export LDFLAGS="-L/opt/homebrew/opt/openssl@3/lib $LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/openssl@3/include $CPPFLAGS"
export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@3/lib/pkgconfig"
****************************

3. Set .zshrc file as below
****************************
. /opt/homebrew/opt/asdf/libexec/asdf.sh
****************************

4. Installed different (older) version just to make sure there are no issues with latest version by running below, which succeeded and did not even try to download openssl this time

asdf install ruby 3.1.1


5. Installed latest version by running below, which succeeded

asdf install ruby 3.2.2

**Reply by mislav** [reply]
Thanks for the update. BTW you only need to set up RUBY_CONFIGURE_OPTS in your environment to link to Homebrew's openssl, but you don't need to edit PATH, LDFLAGS, CPPFLAFS, nor PKG_CONFIG_PATH. You may remove all those settings from your environment, as they will affect all compilation of programs unrelated to Ruby in the future.

**Comment by Mishaaa123** [comment]
oh its hard so
